File: Problemset/0450-Delete Node in a BST/DFS.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} key
# @return {TreeNode}
def delete_node(root, key)
  return nil if root.nil?

  if root.val == key
    root = merge(root.left, root.right)
  elsif root.val > key
    root.left = delete_node(root.left, key)
  else
    root.right = delete_node(root.right, key)
  end

  root
end

# @param {TreeNode} left
# @param {TreeNode} right
# @return {TreeNode}
def merge(left, right)
  return right if left.nil?

  curr = left
  curr = curr.right until curr.right.nil?
  curr.right = right

  left
end

File: Problemset/0450-Delete Node in a BST/DFS.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: TreeNode, key: int) -> TreeNode:
        if root is None:
            return None

        if root.val == key:
            root = self.merge(root.left, root.right)
        elif root.val > key:
            root.left = self.deleteNode(root.left, key)
        else:
            root.right = self.deleteNode(root.right, key)

        return root

    def merge(self, left: TreeNode, right: TreeNode) -> TreeNode:
        if left is None:
            return right

        curr = left
        while curr.right is not None:
            curr = curr.right
        curr.right = right

        return left
